### PR TITLE
multiple work

### PR DESCRIPTION
**Pull request by adeelq**
@cdiacon review this PR first https://github.com/dotmailer/dotmailer-magento2-extension/pull/328

### CODE REVIEW

File: Block/Adminhtml/Config/Automation/Customdatafields.php [hunk 1]
```diff
@@ -12,6 +12,8 @@ class Customdatafields extends
     public $programFactory;
 
     public $elementFactory;
+    private $_getStatusRenderer;
```
Comment (cdiacon) [inline Block/Adminhtml/Config/Automation/Customdatafields.php]: remove underscore

File: Block/Adminhtml/Config/Automation/Customdatafields.php [hunk 2]
```diff
@@ -12,6 +12,8 @@ class Customdatafields extends
...
+    private $_getAutomationRenderer;
```
Comment (cdiacon) [inline Block/Adminhtml/Config/Automation/Customdatafields.php]: remove underscore

File: Controller/Adminhtml/Connector/Ajaxlogcontent.php [hunk 3]
```diff
@@ -10,22 +10,23 @@ class Ajaxlogcontent extends \Magento\Backend\App\Action
...
      *
      * @param \Dotdigitalgroup\Email\Helper\File $file
+     * @param \Magento\Framework\Serialize\Serializer\Json $jsonHelper
      * @param \Magento\Backend\App\Action\Context $context
      * @param \Magento\Framework\Escaper $escaper
      */
     public function __construct(
         \Dotdigitalgroup\Email\Helper\File $file,
-        \Magento\Framework\Json\Helper\Data $jsonHelper,
+        \Magento\Framework\Serialize\Serializer\Json $jsonHelper,
```
Comment (cdiacon) [inline Controller/Adminhtml/Connector/Ajaxlogcontent.php]: change back the class Magento\Framework\Serialize\Serializer\Json was introduced only in 2.2

File: Controller/Adminhtml/Connector/Ajaxvalidation.php [hunk 4]
```diff
@@ -10,19 +10,20 @@ class Ajaxvalidation extends \Magento\Backend\App\Action
...
      * Ajaxvalidation constructor.
      *
      * @param \Dotdigitalgroup\Email\Helper\Data $data
+     * @param \Magento\Framework\Serialize\Serializer\Json $jsonHelper
      * @param \Magento\Backend\App\Action\Context $context
      */
     public function __construct(
         \Dotdigitalgroup\Email\Helper\Data $data,
-        \Magento\Framework\Json\Helper\Data $jsonHelper,
+        \Magento\Framework\Serialize\Serializer\Json $jsonHelper,
```
Comment (cdiacon) [inline Controller/Adminhtml/Connector/Ajaxvalidation.php]: change back the class Magento\Framework\Serialize\Serializer\Json was introduced only in 2.2

File: Controller/Adminhtml/Rules/Ajax.php [hunk 5]
```diff
@@ -40,7 +40,7 @@ class Ajax extends \Magento\Backend\App\AbstractAction
      * @param \Dotdigitalgroup\Email\Model\Adminhtml\Source\Rules\Type      $ruleType
      * @param \Dotdigitalgroup\Email\Model\Adminhtml\Source\Rules\Condition $ruleCondition
      * @param \Dotdigitalgroup\Email\Model\Adminhtml\Source\Rules\Value     $ruleValue
-     * @param \Magento\Framework\Json\Encoder                               $jsonEncoder
+     * @param \Magento\Framework\Serialize\Serializer\Json                  $jsonEncoder
```
Comment (cdiacon) [inline Controller/Adminhtml/Rules/Ajax.php]: class only exists in 2.2

File: Controller/Adminhtml/Rules/Selected.php [hunk 6]
```diff
@@ -42,7 +42,7 @@ class Selected extends \Magento\Backend\App\AbstractAction
      * @param \Dotdigitalgroup\Email\Model\Adminhtml\Source\Rules\Type      $ruleType
      * @param \Dotdigitalgroup\Email\Model\Adminhtml\Source\Rules\Condition $ruleCondition
      * @param \Dotdigitalgroup\Email\Model\Adminhtml\Source\Rules\Value     $ruleValue
-     * @param \Magento\Framework\Json\Encoder                               $jsonEncoder
+     * @param \Magento\Framework\Serialize\Serializer\Json                  $jsonEncoder
```
Comment (cdiacon) [inline Controller/Adminhtml/Rules/Selected.php]: calss only exists in 2.2

File: Controller/Adminhtml/Rules/Value.php [hunk 7]
```diff
@@ -26,14 +26,14 @@ class Value extends \Magento\Backend\App\AbstractAction
      * Value constructor.
      *
      * @param \Dotdigitalgroup\Email\Model\Adminhtml\Source\Rules\Value $ruleValue
-     * @param \Magento\Framework\Json\Encoder                           $jsonEncoder
+     * @param \Magento\Framework\Serialize\Serializer\Json              $jsonEncoder
```
Comment (cdiacon) [inline Controller/Adminhtml/Rules/Value.php]: class only exists in 2.2

File: Controller/Email/Accountcallback.php [hunk 8]
```diff
@@ -31,15 +31,15 @@ class Accountcallback extends \Magento\Framework\App\Action\Action
...
+     * @param \Magento\Framework\Serialize\Serializer\Json            $jsonHelper
      * @param \Magento\Store\Model\StoreManagerInterface              $storeManager
      * @param \Magento\Framework\HTTP\PhpEnvironment\RemoteAddress    $remoteAddress
      * @param \Dotdigitalgroup\Email\Model\Trial\TrialSetup $trialSetup
      */
     public function __construct(
         \Magento\Framework\App\Action\Context $context,
         \Dotdigitalgroup\Email\Helper\Data $helper,
-        \Magento\Framework\Json\Helper\Data $jsonHelper,
+        \Magento\Framework\Serialize\Serializer\Json $jsonHelper,
```
Comment (cdiacon) [inline Controller/Email/Accountcallback.php]: class only exists in 2.2

File: Model/Config/Backend/ArraySerialized.php [hunk 9]
```diff
@@ -5,9 +5,7 @@
 class ArraySerialized extends \Dotdigitalgroup\Email\Model\Config\Backend\Serialized
 {
     /**
-     * Unset array element with '__empty' key
-     *
-     * @return $this
+     * @return \Dotdigitalgroup\Email\Model\Config\Backend\Serialized
```
Comment (cdiacon) [inline Model/Config/Backend/ArraySerialized.php]: keep the comment :
Unset array element with '__empty' key

File: Model/Config/Source/Automation/Program.php [hunk 10]
```diff
@@ -23,6 +23,14 @@ class Program implements \Magento\Framework\Data\OptionSourceInterface
      * @var \Magento\Config\Model\Config\Structure
      */
     public $configStructure;
+    /**
+     * @var \Magento\Store\Model\StoreManagerInterface
+     */
+    private $_storeManager;
```
Comment (cdiacon) [inline Model/Config/Source/Automation/Program.php]: remove underscore

File: Model/Config/Source/Automation/Program.php [hunk 11]
```diff
@@ -23,6 +23,14 @@ class Program implements \Magento\Framework\Data\OptionSourceInterface
...
+    /**
+     * @var \Magento\Framework\Registry
+     */
+    private $_registry;
```
Comment (cdiacon) [inline Model/Config/Source/Automation/Program.php]: remove underscore

File: Model/Config/Source/Datamapping/Datafields.php [hunk 12]
```diff
@@ -20,6 +20,10 @@ class Datafields implements \Magento\Framework\Option\ArrayInterface
      * @var \Magento\Config\Model\Config\Structure
      */
     public $configStructure;
+    /**
+     * @var \Magento\Store\Model\StoreManagerInterface
+     */
+    private $_storeManager;
```
Comment (cdiacon) [inline Model/Config/Source/Datamapping/Datafields.php]: remove underscore

File: Model/Customer/AccountManagement.php [hunk 13]
```diff
@@ -109,7 +108,9 @@ protected function sendEmailConfirmation(\Magento\Customer\Api\Data\CustomerInte
         ) {
             return $this;
         } else {
-            return parent::sendEmailConfirmation($customer, $redirectUrl);
+            parent::sendEmailConfirmation($customer, $redirectUrl);
         }
+
+        return $this;
```
Comment (cdiacon) [inline Model/Customer/AccountManagement.php]: not required  -remove

**Comment by adeelq** [comment]
PHP Fatal error:  Uncaught Error: Call to undefined method Dotdigitalgroup\\Email\\Model\\ResourceModel\\Contact\\Collection::getId() in /var/www/html/m220ce/app/code/Dotdigitalgroup/Email/Observer/Customer/CreateUpdateContact.php:104

**Comment by adeelq** [comment]
Uncaught Error: Call to undefined method Magento\\Wishlist\\Model\\ResourceModel\\Wishlist::getItemCollection() in /var/www/html/m220ce/app/code/Dotdigitalgroup/Email/Observer/Customer/RegisterWishlistItem.php:55

**Comment by TrGlen** [comment]
Create an exclusion rule (type: Review Email Exclusion) generates an exception: Invalid Argument supplied for foreach()...

**Comment by adeelq** [comment]
@TrGlen which site did you get this error. i'm not getting it. so i think you did it on older code.

**Pull request merged**